packages/advanced-logic/specs/payment-context-exchange-rate.md

```diff
@@ -0,0 +1,122 @@
+
+## Description
+
+This extension describes how a request denominated in one currency should be paid in another currency.
+
+When this extension exists, its currency overrides the request currency from a payment point of view.
+The request denomination currency is used to compute the amount in another currency at the payment.
+
+The typical scenario is a request denominated in fiat and paid in ERC20.
+
+Many Exchange rates extensions can be associated with one request, together they describe:
+* To the payment processor what are the possible payment currencies and how to compute corresponding amounts
+* To the payment detection how to compare payments in one currency with a balance in another currency
+
+The version 0.1.0 is not designed to handle properly multiple payments, although it will detect them and
+compute a balance that is very close to the one payers and payee expect.
+
+## Properties
+
+| Property             | Type     | Description                         | Requirement   |
+|----------------------|----------|-------------------------------------|---------------|
+| **id**               | String   | constant value: "pc-exchange-rate"  | **Mandatory** |
+| **type**             | String   | constant value: "paymentContext"    | **Mandatory** |
+| **version**          | String   | constant value: "0.1.0"             | **Mandatory** |
+| **values**           | Object   |                                     |               |
+| **values.oracle**    | String   | TODO:should be fixed for the moment | **Mandatory** |
+| **values.timeframe** | Integer  | Exchange rate timespan              | **Mandatory** |
+| **values.currency**  | Currency | Currency of the expected amount     | **Mandatory** |
+
+---
+
+## Interpretation
+
+When this extension is present at least once, the request payment should be made in one of the exchange rate
+extensions currencies.
+
+### To initiate a payment
+
+The first step for payment initation is to pick one `values.currency` and its related exchange rate extension.
+There should be a payment network for this currency bound to the request.
+
+For the selected currency, the payment processor should look at the exchange rate given by the `values.oracle`.
+Once the exchange rate is fetched, the payment processor has a maximum `values.timeframe`to execute the payment, 
+in seconds. The timeframe will be computed based on the payment transaction datetime, so the payment processor 
+should anticipate network delays by adding a time margin.
+
+### To detect payments
+
+#### Standard case: one payment
+
+The first step is to detect the payment as described in the payment network extensions.
+
+If one payment is detected, we can compute an observed exchange rate, that we can compare to the oracle rate.
+The observed rate should be between the minimum and maximum oracle exchange rate of the period:
+`[paymentDateTime - values.timeframe, paymentDateTime]`.
```

Comment:
What happens in this case:

> I make a payment of 4, since that's the current value of the exchange for me. The min on my timeframe was 1, the max was 10. 
What will be the balance value on the request?
How do we make sure that the balance value is not greater or smaller than the request value (making us show the request as underpaid or overpaid)?

Reply:
The balance is 100% because 4 is between 1 and 10. 
4 is the observed rate, it is between min and max [1, 10] of the timeframe.

Reply:
I'm still not completelly convinced. The amount of the request is an important accounting value for both sides, and mainly for the tool using request. I could clearly see this creating mismatches between different values in an app (e.g.: how did I have 10, pay 4, and am left with 3 on my wallet?)
I wonder if having a magin at the balance object would not be interesting for precision: IBalanceWithEvents could show the precise balance of the request (in our example, 4), and the rate extension could add an event (or something else) that shows the margin of values this amount could represent (in our example, 1 to 10).

Reply:
For me, having the rate timeframe validity in here can make for a breaking UX, if I make a payment and then because of the volatility it shows as underpaid / overpaid this is not an acceptable UX.

Personally, I think we should keep it more simple and just have a timeboxed fixed amount that the user must pay (say 30 minutes) and this value must be regenerated after expiry. (a lot do this with BTC payments).

As this is all handled off-chain we don't have a way of reverting the payment like we could with a smart contract, if we implemented a payment method like this using on-chain oracles with an upper / lower limit it could suffice as we could revert the transaction but potentially having under/over payments for me isn't a good UX.